apollo-router/src/services/layers/content_negociation.rs

```diff
@@ -181,75 +180,48 @@ fn content_type_is_json(headers: &HeaderMap) -> bool {
-/// or if there is no `accept` header
-fn accepts_json(headers: &HeaderMap) -> bool {
-    !headers.contains_key(ACCEPT)
-        || headers.get_all(ACCEPT).iter().any(|value| {
-            value
-                .to_str()
-                .map(|accept_str| {
-                    let mut list = MediaTypeList::new(accept_str);
-
-                    list.any(|mime| {
-                        mime.as_ref()
-                            .map(|mime| {
-                                (mime.ty == APPLICATION && mime.subty == JSON)
-                                    || (mime.ty == APPLICATION
-                                        && mime.subty.as_str() == "graphql-response"
-                                        && mime.suffix == Some(JSON))
-                            })
-                            .unwrap_or(false)
-                    })
-                })
-                .unwrap_or(false)
-        })
-}
-
-/// Returns true if the headers contain header `accept: */*`
-fn accepts_wildcard(headers: &HeaderMap) -> bool {
-    headers.get_all(ACCEPT).iter().any(|value| {
-        value
-            .to_str()
-            .map(|accept_str| {
-                let mut list = MediaTypeList::new(accept_str);
-
-                list.any(|mime| {
-                    mime.as_ref()
-                        .map(|mime| (mime.ty == _STAR && mime.subty == _STAR))
-                        .unwrap_or(false)
-                })
-            })
-            .unwrap_or(false)
-    })
-}
-
-/// Returns true if the headers contain accept header to enable defer
-fn accepts_multipart(headers: &HeaderMap) -> bool {
-    headers.get_all(ACCEPT).iter().any(|value| {
-        value
-            .to_str()
-            .map(|accept_str| {
-                let mut list = MediaTypeList::new(accept_str);
-
-                list.any(|mime| {
-                    mime.as_ref()
-                        .map(|mime| {
-                            mime.ty == MULTIPART
-                                && mime.subty == MIXED
-                                && mime.get_param(
-                                    mediatype::Name::new(MULTIPART_DEFER_SPEC_PARAMETER)
-                                        .expect("valid name"),
-                                ) == Some(
-                                    mediatype::Value::new(MULTIPART_DEFER_SPEC_VALUE)
-                                        .expect("valid value"),
-                                )
-                        })
-                        .unwrap_or(false)
-                })
-            })
-            .unwrap_or(false)
-    })
+// Clippy suggests `for mime in MediaTypeList::new(str).flatten()` but less indentation
+// does not seem worth making it invisible that Result is involved.
+#[allow(clippy::manual_flatten)]
+/// Returns (accepts_json, accepts_wildcard, accepts_multipart)
+fn parse_accept(headers: &HeaderMap) -> (bool, bool, bool) {
```

Comment:
Maybe that would be interesting to return a more explicit type to not have to read the documentation to know which bool is for json and so one. And for subscription we're adding one more case so ... it might be more readable I thin

Reply:
I agree. Let's have a single type that contains the three bools, that will also remove a lot of `Context` access

Reply:
here's the structure I used in another PR: https://github.com/apollographql/router/pull/2802/files#diff-f7b61301740dfc5a5ffa9df3132623479a86d19ae24d1c8c8ac8ee4b63bac3b8R251-R255

Reply:
Thanks, I’ve copied that struct and used it instead of a tuple.

To remove `Context` access, do you mean replacing three the three boolean values in separate keys with that struct in a single key? I wonder if that would be a net positive since derived serialization of the struct will involve a JSON `Value` containing a map with string keys for each struct field.

Reply:
In the other PR I'll replace the context access with the private map, so that will remove the json step anyway